Add CreateAttachment.description and EditAttachment

kangalio opened the pull request:
Summary: support attachment description, replace ad-hoc attachment edit methods with designated EditAttachments builder

Based on #2570

CreateAttachment is now Serialize and can be used for the `attachments` field directly, eliminating the extra `files` fields.

Attachment handling is now centralized in EditAttachments. This removes redundancy in serenity code and makes it more maintainable. Also, by modifying attachments through EditAttachments instead of ad-hoc methods on the builder itself, the side effects of adding attachments can now be documented much more clearly, as well as mirroring those side effects in the builder syntax (also see https://github.com/serenity-rs/serenity/pull/2570#discussion_r1373236965). The old ad-hoc methods are kept for backwards-compatibility, but marked deprecated

Kneemund committed:
Add attachments to payloads, support description

Kneemund committed:
Add separate struct for new message attachments

kangalio committed:
Make everything cleaner (in my judgment, that is)

CreateAttachment is now Serialize and can be used for the `attachments` field directly, eliminating the extra `files` fields.

Attachment handling is now centralized in EditAttachments. This removes redundancy in serenity code and makes it more maintainable. Also, by modifying attachments through EditAttachments instead of ad-hoc methods on the builder itself, the side effects of adding attachments can now be documented much more clearly, as well as mirroring those side effects in the builder syntax (also see https://github.com/serenity-rs/serenity/pull/2570#discussion_r1373236965). The old ad-hoc methods are kept for backwards-compatibility, but marked deprecated

mkrasnitski commented on src/builder/create_attachment.rs:
```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+
+    /// This method removes an existing attachment from the list of attachments that are kept after
+    /// editing.
+    pub fn dont_keep(mut self, id: AttachmentId) -> Self {
```
I think a better name would be `remove`, or maybe `drop`.

mkrasnitski replied:
Any objections to changing this?

kangalio replied:
Your envisioned mental model of EditAttachments seems to be: two seperate lists of attachments, initially empty, that each can be pushed to. This more closely matches internals than my envisioned mental model of EditAttachments, which is: one list of new attachments that is initially empty and can be pushed to; and a set of all existing attachments, initially full but with all items set "inactive", where you don't push to the set but rather activate specific attachments by their ID (think `existing_attachments.set_to_keep(id, true);` rather than `existing_attachments.add_to_existing(id);`)

My idea is that the latter mental model is more intuitive to people who don't happen to already know how the API works under the hood. And that's also why I chose the terms "keep"/"dont_keep"

mkrasnitski replied:
I think the name `remove` still fits with this mental model, as it marks the existing attachment to be explicitly "removed" or otherwise "excluded" from the message. And I'd say that I do in fact view the existing attachments as a set and not as an empty list. I just really dislike the use of `dont` as part of the method name like this. It's normally a contraction but obviously you can't put an apostrophe in a method name (I'm not sure if that would make it a worse name or better tbh lol). As it stands it reads not seriously at all.

In principle, the pair of `keep`/`dont_keep` sounds too passive for me. If it were `keep`/`remove`, that would be better. The pair `keep`/`drop` has a nice symmetry, but `drop` is not a very conventional term, so I like it less. There may be some replacement for `keep` that's also less passive, but I can't think of one at the moment.

mkrasnitski commented on src/http/multipart.rs:
```diff
@@ -5,11 +5,23 @@ use reqwest::multipart::{Form, Part};
+        let mut part = Part::bytes(self.data);
+        part = guess_mime_str(part, &self.filename)?;
+        part = part.file_name(self.filename);
+        Ok(part)
+    }
+}
+
 /// Holder for multipart body. Contains files, multipart fields, and payload_json for creating
 /// requests with attachments.
 #[derive(Clone, Debug)]
 pub struct Multipart {
-    pub files: Vec<CreateAttachment>,
+    /// Files that are sent with the form data as individual uploads.
+    pub upload_file: Option<CreateAttachment>,
+    /// Files that are sent with the form data as message attachments.
+    pub attachment_files: Option<Vec<CreateAttachment>>,
```
Do we still want to combine these into a single enum, since they seem to be mutually exclusive? Would remove the need for `CreateAttachment::into_part` as the part could be calculated inline using a `match`.

kangalio replied:
Sure we can, I don't particularly care. I frankly forgot that idea was a thing while creating this follow-up PR

mkrasnitski replied:
I experimented a bit with names locally, and I kind of like this definition:
```rust
pub enum MultipartUpload {
    /// A file sent with the form data as an individual upload. For example, a sticker.
    File(CreateAttachment),
    /// Files sent with the form as message attachments.
    Attachments(Vec<CreateAttachment>),
}
```
Originally, if you remember, the proposed enum name was simply `Files`, but that seemed too generic to me. And, on its own the `File` variant may also seem generic but when written as `MultipartUpload::File`, I think it's pretty understandable. It only would occur in a single place anyways.

kangalio replied:
I like it too! Nice idea

mkrasnitski commented on src/builder/create_attachment.rs:
```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+
+    /// This method adds an existing attachment to the list of attachments that are kept after
+    /// editing.
+    pub fn keep(mut self, id: AttachmentId) -> Self {
```
What do you think of the name `retain`?

kangalio replied:
Not a fan tbh. In Rust I associate it with `collection.retain(<closure>)`, which this is not. Also, I'm always a fan if method names match natural language, and I'd naturally say "keep an attachment" rather than "retain an attachment" (for example in the docs of that method)

mkrasnitski commented on src/builder/create_attachment.rs:
```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+    ///
+    /// **Note: this EditAttachments must be run on the same message as is supplied here, or else
+    /// Discord will throw an error!**
+    pub fn new_keep_all(msg: &Message) -> Self {
```
The `new_` prefix is unnecessary. This could also be named `retain_all`, similar to `keep` above. What do you think?

kangalio replied:
I went through a few iterations on the method name and deliberately decided to keep `new`, to make clear that this is a constructor method.

mkrasnitski replied:
That's true, but the signature is pretty short and the absence of `self` I think is easy enough to notice. Moreover, I think having the name be consistent with `keep` is important.

kangalio replied:
Alright

mkrasnitski commented on src/builder/create_attachment.rs:
```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+    /// Clones all new attachments into a new Vec, keeping only data and filename, because those
+    /// are needed for the multipart form data. The data is taken out of `self` in the process, so
+    /// this method can only be called once.
+    pub(crate) fn take_files(&mut self) -> Vec<CreateAttachment> {
```
Could be renamed to just `take`, though not that important.

kangalio replied:
I prefer `take_files` over `take`, but I'll not expand on that if it's not that important

mkrasnitski commented on src/builder/create_attachment.rs:
```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+    /// are needed for the multipart form data. The data is taken out of `self` in the process, so
+    /// this method can only be called once.
+    pub(crate) fn take_files(&mut self) -> Vec<CreateAttachment> {
+        let mut id_placeholder = 0;
```
This variable would be better named `index` or something related. Ideally you could just call `.filter_map().enumerate()` on the iterator instead of manually incrementing a counter.

kangalio replied:
I deliberately did not name it `index` because that's semantically not what it is, as intended by Discord. It really is a placeholder for the ID, because Discord hasn't assigned yet. That's the "click" moment I talked about earlier in the Discord. I haven't tested, but I'm pretty sure the ID placeholder could be any numbers, not just sequential starting from 0

mkrasnitski replied:
Are you saying that the id doesn't have to match the `files[N]` name of the new attachment?

kangalio replied:
No, they do have to match, but I believe it doesn't have to be 0, 1, 2; it can be 42, 69, 420, 1337. So `files[1337]` in the multipart and `{"id": "1337"}` in the payload_json

mkrasnitski commented on src/builder/create_attachment.rs:
```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+
+    /// Adds a new attachment to the attachment list.
+    #[allow(clippy::should_implement_trait)] // Clippy thinks add == std::ops::Add::add
+    pub fn add(mut self, attachment: CreateAttachment) -> Self {
```
Can't think of an alternative name, since `push` doesn't really make sense. Willing to discuss further though.

kangalio replied:
Why an alternative name? I like `add` very much - short, simple, self-explanatory. The fact that clippy throws a fit is not a reason to discard that name in my opinion

mkrasnitski replied:
Mainly because it's slightly ambiguous what you're adding if you only look at the method name, if it's existing or new attachments.

kangalio replied:
In another comment I explained my envisioned mental model here. The idea is that existing attachments shouldn't be thought of as a pushable list, but a set with togglable items. Hence, semantically, the only pushable list is the list of new attachments and `add` is unambiguous

kangalio committed:
new_keep_all -> keep_all as per mkrasnitski/polarbits' suggestion

mkrasnitski commented on src/builder/create_attachment.rs:
```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+/// When this builder is _not_ supplied in a message edit, Discord keeps the attachments intact.
+/// However, as soon as a builder is supplied, Discord removes all attachments from the message.
+/// If you want to keep old attachments, you have to specify this for each attachment using
+/// [`Self.keep()`].
```
```suggestion
/// However, as soon as a builder is supplied, Discord removes all attachments from the message. If
/// you want to keep old attachments, you must specify this either using [`Self::keep_all`], or
/// individually for each attachment using [`Self::keep`].
```

mkrasnitski commented on src/builder/create_attachment.rs:
```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+    /// Creates a new attachments builder that keeps all existing attachments.
+    ///
+    /// Shorthand for [`Self::new()`] and calling [`Self::keep()`] for every
+    /// [`AttachmentId`] in [`Message::attachments`].
```
```suggestion
    /// Shorthand for [`Self::new()`] and calling [`Self::keep()`] for every [`AttachmentId`] in
    /// [`Message::attachments`].
```

mkrasnitski commented on examples/testing/src/main.rs:
```diff
@@ -291,8 +291,9 @@ async fn interaction(
                 ctx,
-                EditInteractionResponse::new()
-                    .new_attachment(CreateAttachment::url(ctx, IMAGE_URL_2).await?),
+                EditInteractionResponse::new().attachments(
+                    EditAttachments::new().add(CreateAttachment::url(ctx, IMAGE_URL_2).await?),
+                ),
```
This is a bit wordy. Do you think some of the now-deprecated builder methods could instead be turned into wrappers around `.attachments`? That way, complexity is still centralized into `EditAttachments` but convenience methods are still left. I'd nominate `new_attachment` be kept at least.

kangalio replied:
Good point. Maybe I can actually just un-deprecate all those methods, while carefully documenting what `EditAttachments` invocation they're equivalent to, so users can read about and grasp the unintuitive side effects of those methods.

Because the existing methods were already chosen carefully a few months ago to minimize side effect confusion while not devolving into wordiness, so their design is probably exactly what we want for "less wordier shorthands for EditAttachments"

mkrasnitski replied:
This should be reverted now that `new_attachment` is no longer deprecated.

kangalio committed:
Update src/builder/create_attachment.rs

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

kangalio committed:
Update src/builder/create_attachment.rs

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

kangalio committed:
Rename dont_keep to remove

As per the principle of "do what you can be blamed the least for", even though I still don't like this rename

kangalio committed:
Add MultipartUpload enum

kangalio committed:
Undeprecate methods

mkrasnitski commented on src/http/client.rs:
```diff
@@ -26,6 +26,7 @@ use super::{
     HttpError,
     LightMethod,
     MessagePagination,
+    MultipartUpload,
```
This should be imported directly along with `super::multipart::Multipart`, a few lines above this.

mkrasnitski commented on src/http/client.rs:
```diff
@@ -489,8 +490,7 @@ impl Http {
             request.multipart = Some(Multipart {
-                attachment_files: Some(files.into_iter().map(Into::into).collect()),
-                upload_file: None,
+                upload: MultipartUpload::Attachments(files.into_iter().map(Into::into).collect()),
```
The `.into_iter().map(Into::into).collect()` is actually a no-op because `files` is already a `Vec<CreateAttachment>`. I won't duplicate this note for every other instance, but in all cases they could just take `files` directly.

kangalio replied:
Lol

mkrasnitski commented on src/http/multipart.rs:
```diff
@@ -14,14 +14,19 @@ impl CreateAttachment {
-    pub upload_file: Option<CreateAttachment>,
-    /// Files that are sent with the form data as message attachments.
-    pub attachment_files: Option<Vec<CreateAttachment>>,
+    pub upload: MultipartUpload,
```
Should this still be called `files`? If not, we might want to adjust the docstring on `Multipart` somehow.

kangalio committed:
Apply mkrasnitski/polarbits' suggestions

mkrasnitski commented on src/http/multipart.rs:
```diff
@@ -5,11 +5,28 @@ use reqwest::multipart::{Form, Part};
+    Attachments(Vec<CreateAttachment>),
+}
+
+/// Holder for multipart body. Contains upload data, multipart fields, and payload_json for creating
 /// requests with attachments.
```
```suggestion
/// Holder for multipart body. Contains upload data, multipart fields, and payload_json for
/// creating requests with attachments.
```
Sorry for the nit

mkrasnitski commented:
It seems you undeprecated the old methods without actually changing them to wrap `EditAttachments` internally?

kangalio commented:
They already used EditAttachments internally, no? I changed it in the initial commit already

mkrasnitski commented:
~~As far as I can tell, the implementations on `EditMessage` and `EditWebhookMessage` are unchanged.~~ Also, the methods on `EditInteractionResponse` still contain `#[allow(deprecated)]` even though the wrapped methods are no longer deprecated.

EDIT: Nevermind, I see what you mean. I hadn't realized that they were already expressed in terms of `EditAttachments`.

mkrasnitski commented on src/builder/edit_interaction_response.rs:
```diff
@@ -75,29 +76,30 @@ impl EditInteractionResponse {
     ///
-    /// If this is called one or more times, existing attachments will reset. To keep them, provide
-    /// their IDs to [`Self::keep_existing_attachment`].
+    /// Resets existing attachments. See [`Self::keep_existing_attachment`] or read
+    /// [`EditAttachments`] for explanation.
     pub fn new_attachment(self, attachment: CreateAttachment) -> Self {
+        #[allow(deprecated)]
```
```suggestion
    /// Resets existing attachments. See the documentation for [`EditAttachments`] for details.
    pub fn new_attachment(self, attachment: CreateAttachment) -> Self {
```

mkrasnitski commented on src/builder/edit_interaction_response.rs:
```diff
@@ -75,29 +76,30 @@ impl EditInteractionResponse {
-    /// Keeps an existing attachment by id.
-    ///
-    /// To be used after [`Self::new_attachment`] or [`Self::clear_existing_attachments`].
+    /// Shorthand for [`EditAttachments::keep`]
     pub fn keep_existing_attachment(self, id: AttachmentId) -> Self {
+        #[allow(deprecated)]
```
```suggestion
    /// Shorthand for [`EditAttachments::keep`].
    pub fn keep_existing_attachment(self, id: AttachmentId) -> Self {
```

mkrasnitski commented on src/builder/edit_interaction_response.rs:
```diff
@@ -75,29 +76,30 @@ impl EditInteractionResponse {
-    /// keep only some existing attachments.
-    pub fn clear_existing_attachments(self) -> Self {
-        Self(self.0.clear_existing_attachments())
+    /// Shorthand for [`Self::attachments`] with [`EditAttachments::new`]
+    pub fn clear_attachments(self) -> Self {
+        #[allow(deprecated)]
```
```suggestion
    /// Shorthand for calling [`Self::attachments`] with [`EditAttachments::new`].
    pub fn clear_attachments(self) -> Self {
```

mkrasnitski commented on src/builder/edit_message.rs:
```diff
@@ -190,35 +187,40 @@ impl EditMessage {
+    /// Adds a new attachment to the message.
     ///
-    /// This can be called multiple times.
+    /// Resets existing attachments. See [`Self::keep_existing_attachment`] or read
+    /// [`EditAttachments`] for explanation.
     pub fn attachment(mut self, attachment: CreateAttachment) -> Self {
```
```suggestion
    /// Resets existing attachments. See the documentation for [`EditAttachments`] for details.
    pub fn new_attachment(mut self, attachment: CreateAttachment) -> Self {
```
This method name isn't in line with the `new_attachment` method on `EditWebhookMessage`/`EditInteractionResponse`, and in combination with the new `attachments` method, might be confusing. I think it'd be better to change the name.

mkrasnitski commented on src/builder/edit_message.rs:
```diff
@@ -190,35 +187,40 @@ impl EditMessage {
     }
 
-    /// Remove all attachments.
+    /// Shorthand for [`Self::attachments`] with [`EditAttachments::new`]
```
```suggestion
    /// Shorthand for calling [`Self::attachments`] with [`EditAttachments::new`].
```

mkrasnitski commented on src/builder/edit_webhook_message.rs:
```diff
@@ -133,33 +131,32 @@ impl EditWebhookMessage {
     ///
-    /// If this is called one or more times, existing attachments will reset. To keep them, provide
-    /// their IDs to [`Self::keep_existing_attachment`].
+    /// Resets existing attachments. See [`Self::keep_existing_attachment`] or read
+    /// [`EditAttachments`] for explanation.
```
```suggestion
    /// Resets existing attachments. See the documentation for [`EditAttachments`] for details.
```

mkrasnitski commented on src/builder/edit_webhook_message.rs:
```diff
@@ -133,33 +131,32 @@ impl EditWebhookMessage {
-    /// keep only some existing attachments.
-    pub fn clear_existing_attachments(mut self) -> Self {
-        self.attachments = Some(Vec::new());
+    /// Shorthand for [`Self::attachments`] with [`EditAttachments::new`]
```
```suggestion
    /// Shorthand for calling [`Self::attachments`] with [`EditAttachments::new`].
```

mkrasnitski commented on src/builder/edit_message.rs:
```diff
@@ -190,35 +187,40 @@ impl EditMessage {
     }
 
-    /// Add an existing attachment by id.
+    /// Shorthand for [`EditAttachments::keep`]
     pub fn add_existing_attachment(mut self, id: AttachmentId) -> Self {
```
```suggestion
    /// Shorthand for [`EditAttachments::keep`].
    pub fn keep_existing_attachment(mut self, id: AttachmentId) -> Self {
```
Same here as for `EditMessage::attachment`. The method name should be inline with `EditWebhookMessage`.

mkrasnitski commented on src/builder/edit_message.rs:
```diff
@@ -190,35 +187,40 @@ impl EditMessage {
     }
 
-    /// Remove an existing attachment by id.
+    /// Shorthand for [`EditAttachments::remove`]
     pub fn remove_existing_attachment(mut self, id: AttachmentId) -> Self {
```
```suggestion
    /// Shorthand for [`EditAttachments::remove`].
    pub fn remove_existing_attachment(mut self, id: AttachmentId) -> Self {
```
I've noticed there is no equivalent helper method on `EditWebhookMessage`. Maybe worth adding?

kangalio replied:
I suppose. Serenity current doesn't have it so I don't really care. Can always be added later

mkrasnitski commented on src/builder/edit_webhook_message.rs:
```diff
@@ -133,33 +131,32 @@ impl EditWebhookMessage {
-    /// Keeps an existing attachment by id.
-    ///
-    /// To be used after [`Self::new_attachment`] or [`Self::clear_existing_attachments`].
+    /// Shorthand for [`EditAttachments::keep`]
```
```suggestion
    /// Shorthand for [`EditAttachments::keep`].
```

mkrasnitski reviewed (changes requested):
Last round of review (hopefully). Sorry the review is a bit out of order.

kangalio committed:
Update src/builder/edit_interaction_response.rs

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

kangalio committed:
Update src/builder/edit_interaction_response.rs

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

kangalio committed:
Apply suggestions from code review

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

mkrasnitski commented on src/builder/create_attachment.rs:
````diff
@@ -117,4 +106,178 @@ impl CreateAttachment {
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message) -> Result<(), Error> {
````
```suggestion
/// # async fn _foo(ctx: Http, mut msg: Message) -> Result<(), Error> {
```
This fixes the CI failure with no gateway. The other doctests should also be changed in this way.

mkrasnitski commented on src/builder/create_attachment.rs:
```diff
@@ -117,4 +106,178 @@ impl CreateAttachment {
+        files
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
```
```suggestion
    #[cfg(feature = "cache")]
    pub(crate) fn is_empty(&self) -> bool {
```
Get an unused warning if cache is disabled.

mkrasnitski reviewed (changes requested):
Ok, for real this time. Thank you for being so patient.

kangalio committed:
Update src/builder/create_attachment.rs

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

kangalio committed:
Update src/builder/create_attachment.rs

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

mkrasnitski committed:
Fix CI

mkrasnitski reviewed (approved):
Thank you for sticking with this. This PR is good to go.

mkrasnitski commented:
To summarize the breaking changes in this PR (for changelog purposes):
* Renamed the following methods:
    * `{EditInteractionResponse, EditWebhookMessage}::clear_existing_attachments` -> `clear_attachments`
    * `EditMessage::attachment` -> `new_attachment`
    * `EditMessage::add_existing_attachment` -> `keep_existing_attachment`
 * Replaced `Multipart::files` field with an `upload` field of type `MultipartUpload`.
